SPIRE-147: Adds route to oidc-discovery-provider-service

PillaiManish opened the pull request:
This PR focus on adding route on oidc-discovery-provider-service

- Adds annotations service.beta.openshift.io/serving-cert-secret-name: oidc-serving-cert
  - run "make update bin-data"
- Cherry-pick from https://github.com/openshift/zero-trust-workload-identity-manager/pull/28
- Adds route to schema
- Adds generateOIDCDiscoveryProviderRoute for creating route resource

PillaiManish commented:
/cc @anirudhAgniRedhat @TrilokGeer

anirudhAgniRedhat commented on pkg/controller/utils/utils.go:
```diff
@@ -3,6 +3,7 @@ package utils
 import (
 	"crypto/sha256"
 	"encoding/hex"
+	routev1 "github.com/openshift/api/route/v1"
```
nit: can we keep imports sorted in groups as it is there in other files

PillaiManish replied:
Done, PTAL.

anirudhAgniRedhat commented on pkg/controller/spire-oidc-discovery-provider/route.go:
```diff
@@ -0,0 +1,55 @@
+		"app.kubernetes.io/name":     "spiffe-oidc-discovery-provider",
+		"app.kubernetes.io/instance": "spire",
+		"app.kubernetes.io/part-of":  "zero-trust-workload-identity-manager",
+		"app.kubernetes.io/version":  "1.12.0",
```
nits can we use const for this as well?? It would reduce efforts in updating this moving forward.

PillaiManish replied:
Let's fix this in follow-up PR, so that we can replace it in all files at once.

anirudhAgniRedhat commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -347,3 +406,11 @@ func needsUpdate(current, desired appsv1.Deployment) bool {
 }
+
+// needsRouteUpdate returns true if Route needs to be updated
+func needsRouteUpdate(current, desired routev1.Route) bool {
```
can we have tescases for this??

PillaiManish commented:
/retest

PillaiManish commented on pkg/controller/spire-oidc-discovery-provider/routes.go:
```diff
@@ -0,0 +1,56 @@
+// generateOIDCDiscoveryProviderRoute creates an OpenShift Route resource for the SPIRE OIDC Discovery Provider
+func generateOIDCDiscoveryProviderRoute(config *v1alpha1.SpireOIDCDiscoveryProvider) *routev1.Route {
+	labels := map[string]string{
+		"app.kubernetes.io/name":     "spiffe-oidc-discovery-provider",
```
https://github.com/PillaiManish/zero-trust-workload-identity-manager/blob/0123afebe09a6ef4697af0e964b8fd670ed2cbc7/pkg/controller/spire-agent/daemonset.go#L17-L20

https://github.com/PillaiManish/zero-trust-workload-identity-manager/blob/ec66a230f6239b2b2812acaaaf3fb6f2ccc20447/pkg/controller/spiffe-csi-driver/daemonset.go

@lunarwhite should I keep some prefix like this, or keep same as it is?

lunarwhite replied:
Oh, I wasn't aware of these differences you've posted, thanks.

Then we keep same as what EP documented: https://github.com/openshift/enhancements/blob/7cdf7cf7e69cc7e413f97ba854883b2618d17a41/enhancements/workload-identity-management/oidc-routes-integration.md?plain=1#L138, to use `spiffe-oidc-discovery-provider` as name and label.

It also aligns with upstream: https://github.com/spiffe/helm-charts-hardened/blob/86f0aecc5729a68eb6b6aea51c547e5c1236a23b/charts/spire/charts/spiffe-oidc-discovery-provider/templates/ingress.yaml#L39

PillaiManish replied:
Okay, let me change it, thanks @lunarwhite .

PillaiManish commented on pkg/controller/spire-oidc-discovery-provider/routes.go:
```diff
@@ -0,0 +1,56 @@
+			Port: &routev1.RoutePort{
+				TargetPort: intstr.FromString("https"),
+			},
+			TLS: &routev1.TLSConfig{
+				Termination: routev1.TLSTerminationReencrypt,
+			},
```
added.

PillaiManish commented:
Hey @anirudhAgniRedhat, @lunarwhite as discussed I have added the `managedRoute` api & also removed the spiffe-helper from oidc-provider deployment, PTAL.

PillaiManish commented on pkg/controller/spire-oidc-discovery-provider/configmaps.go:
```diff
@@ -61,7 +66,7 @@ func GenerateOIDCConfigMapFromCR(cr *v1alpha1.SpireOIDCDiscoveryProvider) (*core
 	}
 
 	spiffeHelperConf := `agent_address = "/spiffe-workload-api/` + agentSocketName + `"
```
Removed.

lunarwhite reviewed (commented):
/label qe-approved
/verified by premerge

lunarwhite commented:
Just a note: this PR might need to rebase once https://github.com/openshift/zero-trust-workload-identity-manager/pull/28 is landed, if so https://github.com/openshift/zero-trust-workload-identity-manager/pull/33/commits/773e6ac8dfb35560f19eab8bae1f5335a08cfa76 is not required.

anirudhAgniRedhat commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -290,6 +292,58 @@ func (r *SpireOidcDiscoveryProviderReconciler) Reconcile(ctx context.Context, re
+			Name:      route.Name,
+			Namespace: route.Namespace,
+		}, &existingRoute)
+		if err != nil && kerrors.IsNotFound(err) {
```
https://github.com/openshift/enhancements/pull/1824/files#diff-a59a8350084f46459b43dd4c3474f41f94484eb4ff4832a2e9df7daf82312533R46

We need to add this check!!

ldhananj commented on bundle/manifests/operator.openshift.io_spireservers.yaml:
```diff
@@ -1058,7 +1058,7 @@ spec:
                 type: object
               jwtIssuer:
-                description: jwtIssuer is the JWT issuer domain. Defaults to oidc-discovery.$trustDomain
+                description: jwtIssuer is the JWT issuer domain. Defaults to https://oidc-discovery.$trustDomain
```
https://github.com/openshift/zero-trust-workload-identity-manager/pull/28/files implementation is not consider a default jwtIssuer and treated it as a mandatory field. However, this change introduces https://oidc-discovery.$trustDomain as the default value for jwtIssuer

ldhananj commented on config/crd/bases/operator.openshift.io_spireoidcdiscoveryproviders.yaml:
```diff
@@ -979,6 +979,15 @@ spec:
                 type: object
                 x-kubernetes-map-type: granular
+              managedRoute:
+                default: "true"
```
clearly explains what each option true / false enables:

"true": Allows automatic exposure of OIDC discovery endpoints through a managed OpenShift Route (*.apps.<cluster-domain>). 
"false": Allows administrators to manually configure exposure using custom OpenShift Routes or ingress, offering more control over routing behavior.

ldhananj commented on pkg/controller/spire-oidc-discovery-provider/configmaps.go:
```diff
@@ -46,8 +51,8 @@ func GenerateOIDCConfigMapFromCR(cr *v1alpha1.SpireOIDCDiscoveryProvider) (*core
 		"log_level": "debug",
 		"serving_cert_file": map[string]string{
 			"addr":           ":8443",
```
Avoid hardcoding constants in the code. If the value is subject to change, consider retrieving it from a ConfigMap or a centralized constants file. In most cases, referencing it from a ConfigMap is preferable for easier configuration management and environment-specific overrides.

ldhananj commented on pkg/controller/spire-oidc-discovery-provider/configmaps.go:
```diff
@@ -46,8 +51,8 @@ func GenerateOIDCConfigMapFromCR(cr *v1alpha1.SpireOIDCDiscoveryProvider) (*core
+			"key_file_path":  "/etc/oidc/tls/tls.key",
 		},
 		"workload_api": map[string]string{
 			"socket_path":  "/spiffe-workload-api/" + agentSocketName,
```
Prefer using named constants instead of hardcoded values in the code to improve readability, maintainability, and reduce the risk of errors.

ldhananj commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -290,6 +292,58 @@ func (r *SpireOidcDiscoveryProviderReconciler) Reconcile(ctx context.Context, re
 		Message: "Spire OIDC Deployment created",
 	}
+
+	if utils.StringToBool(oidcDiscoveryProviderConfig.Spec.ManagedRoute) {
```
Consider breaking this logic into smaller helper functions like:
setControllerReferenceForRoute(...)
createOrGetRoute(...)
updateReconcileStatus(...)
This improves readability and makes the reconciliation loop easier to maintain.

ldhananj replied:
Ensure this logic is covered in unit tests:

Route creation success
Route already exists
Controller reference failure
Route creation failure
Route retrieval failure

ldhananj commented on pkg/controller/spire-oidc-discovery-provider/deployments.go:
```diff
@@ -93,12 +80,12 @@ func buildDeployment(config *v1alpha1.SpireOIDCDiscoveryProvider, spireOidcConfi
+							Name: "tls-certs",
+							VolumeSource: corev1.VolumeSource{
+								Secret: &corev1.SecretVolumeSource{
+									SecretName: "oidc-serving-cert",
```
Avoid using hardcoded values in the code. Instead, store them in a ConfigMap and retrieve them dynamically to improve flexibility, maintainability, and environment-specific configuration.

TrilokGeer commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -34,6 +35,7 @@ const (
 	SpireOIDCConfigMapGeneration   = "SpireOIDCConfigMapGeneration"
 	SpireOIDCSCCGeneration         = "SpireOIDCSCCGeneration"
 	SpireClusterSpiffeIDGeneration = "SpireClusterSpiffeIDGeneration"
+	ManagedRouteGeneration         = "ManagedRouteGeneration"
```
```suggestion
	ManagedRouteReady         = "ManagedRouteReady"
```

anirudhAgniRedhat commented on pkg/controller/spire-oidc-discovery-provider/routes.go:
```diff
@@ -0,0 +1,57 @@
+		Spec: routev1.RouteSpec{
+			Host: host,
+			Port: &routev1.RoutePort{
+				TargetPort: intstr.FromString("https"),
```
add port check

anirudhAgniRedhat commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -290,6 +292,69 @@ func (r *SpireOidcDiscoveryProviderReconciler) Reconcile(ctx context.Context, re
+			}
+			r.log.Info("Created route", "Namespace", route.Namespace, "Name", route.Name)
+		} else if err == nil && checkRouteConflict(&existingRoute, route) {
+			if err = r.ctrlClient.Update(ctx, &existingRoute); err != nil {
```
we are not updating the route

anirudhAgniRedhat commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -290,6 +292,69 @@ func (r *SpireOidcDiscoveryProviderReconciler) Reconcile(ctx context.Context, re
+			r.log.Info("Updated spire oidc discovery provider deployment")
+		} else {
+			r.log.Error(err, "Failed to get existing route")
+			reconcileStatus[ManagedRouteReady] = reconcilerStatus{
```
if error is not nil and no conflict condition is missing

anirudhAgniRedhat commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -290,6 +292,69 @@ func (r *SpireOidcDiscoveryProviderReconciler) Reconcile(ctx context.Context, re
+			r.log.Error(err, "Failed to get existing route")
+			reconcileStatus[ManagedRouteReady] = reconcilerStatus{
+				Status:  metav1.ConditionFalse,
+				Reason:  "SpireOIDCManagedRouteRetrievalFailed",
```
SpireOIDCManagedRouteRetrievalFailed > ManagedRouteRetrievalFailed

TrilokGeer commented on bundle/manifests/operator.openshift.io_spireoidcdiscoveryproviders.yaml:
```diff
@@ -1177,6 +1179,16 @@ spec:
                 x-kubernetes-list-type: map
+              managedRouteStatus:
+                properties:
+                  jwtIssuer:
```
Update the `jwtIssuer` to `hostname`, as there is no context of jwt issuer in route.

PillaiManish replied:
Done

PillaiManish commented on pkg/controller/spire-oidc-discovery-provider/controller.go:
```diff
@@ -367,3 +374,95 @@ func needsUpdate(current, desired appsv1.Deployment) bool {
 }
+
+// checkRouteConflict returns true if desired & current routes has conflicts else return false
+func checkRouteConflict(current, desired *routev1.Route) bool {
+	if current.Spec.TLS == nil ||
+		current.Spec.TLS.Termination != desired.Spec.TLS.Termination ||
+		current.Spec.TLS.ExternalCertificate != desired.Spec.TLS.ExternalCertificate ||
+		current.Spec.Host != desired.Spec.Host ||
+		current.Spec.To.Name != desired.Spec.To.Name ||
+		current.Spec.To.Kind != desired.Spec.To.Kind ||
+		current.Spec.Port == nil ||
+		current.Spec.Port.TargetPort != desired.Spec.Port.TargetPort {
+		return true
+	}
```
Replace it with deepcopy().

PillaiManish commented on pkg/controller/spire-server/statefulset.go:
```diff
@@ -131,7 +131,7 @@ func GenerateSpireServerStatefulSet(config *v1alpha1.SpireServerSpec,
 			VolumeClaimTemplates: []corev1.PersistentVolumeClaim{
 				{
-					ObjectMeta: metav1.ObjectMeta{Name: "spire-data", Labels: labels},
+					ObjectMeta: metav1.ObjectMeta{Name: "spire-data"},
```
@lunarwhite, this label has to be removed, else on update it throws an error as `Statefulset doesn't allow to update templates`.

PillaiManish commented:
/remove-label tide/merge-method-squash

lunarwhite commented:
In order to use `.spec.tls.externalCertificate` in Route, the following RBACs are required ([ref](https://docs.redhat.com/en/documentation/openshift_container_platform/4.19/html-single/ingress_and_load_balancing/index#nw-ingress-route-secret-load-external-cert_secured-routes)):

```
oc create role secret-reader \
  --verb=get,list,watch \
  --resource=secrets \
  --resource-name=$TLS_SECRET_NAME \
  -n zero-trust-workload-identity-manager

oc create rolebinding secret-reader-binding \
  --role=secret-reader \
  --serviceaccount=openshift-ingress:router \
  -n zero-trust-workload-identity-manager
```

Would it be a good idea to let our operator also create them for users when setting `.spec.externalSecretRef`? Note: It's not a blocking issue, we could take this into consideration when moving towards GA.

lunarwhite commented:
/verified by premerge
